### PR TITLE
Restart broken Olm sessions

### PR DESCRIPTION
**Pull request by dbkr**
unwedge

* Start a new Olm sessions with a device when we get an undecryptable
   message on it.
 * Send a dummy message on that sessions such that the other end knows
   about it.
 * Re-send any outstanding keyshare requests for that device.

Also includes a unit test for megolm that isn't very related but came
out as a result anyway.

Includes #776
Fixes vector-im/riot-web#3822

### CODE REVIEW

**Comment by uhoreg** [comment]
We'll need to make an MSC to document this.  So far, it looks like we'll need to document:

- when an olm session is created and has not yet received a message, use the creation time of the session as the time that it last received a message when deciding what session to use for encrypting outgoing messages
- for last received time of messages, only consider messages that were successfully decrypted
- when you receive an undecryptable message using olm, create a new session and send a dummy event using it so that the other side knows about it (except that you don't have to create a new session if you already did)
- ...

File: spec/unit/crypto.spec.js [hunk 1]
```diff
@@ -1,9 +1,13 @@
-
-"use strict";
```
[the remark this answers is not in the document]
Comment (dbkr) [reply spec/unit/crypto.spec.js]: these are a throwback from a time before modular ES6 so they should generally be removed throughout

File: src/crypto/index.js [hunk 2]
```diff
@@ -1162,6 +1164,74 @@ Crypto.prototype._onRoomKeyEvent = function(event) {
     alg.onRoomKeyEvent(event);
 };
 
+/**
+ * Handle a toDevice event that couldn't be decrypted
+ *
+ * @private
+ * @param {module:models/event.MatrixEvent} event undecryptable event
+ */
+Crypto.prototype._onToDeviceBadEncrypted = async function(event) {
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/crypto/index.js]: right, so the problem with this is that only prekey messages contain information on what session ID / device key they're destined for: normal messages sent once the session has been established do not, so if we have multiple sessions, we won't know which has been replaced (when we get an incoming message, we just try decrypting it with every session we have).

[the remark this answers is not in the document]
Comment (dbkr) [reply src/crypto/index.js]: I can't think of anything better to do than a dumb heuristic like only creating one new session every hour.

[the remark this answers is not in the document]
Comment (uhoreg) [reply src/crypto/index.js]: That's unfortunate.  It seems like using a dumb heuristic might be the only thing to do, though every hour might be too long.  Maybe something more like 10 minutes would be better, as hopefully by then, the sender would have received the m.dummy message.

File: src/crypto/index.js [hunk 3]
```diff
@@ -1180,6 +1191,19 @@ Crypto.prototype._onToDeviceBadEncrypted = async function(event) {
...
+    // check when we last forced a new session with this device: if we've already done so
+    // recently, don't do it again.
+    this._lastNewSessionForced[sender] = this._lastNewSessionForced[sender] || {};
+    const lastNewSessionForced = this._lastNewSessionForced[sender][deviceKey] || 0;
+    if (lastNewSessionForced + MIN_FORCE_SESSION_INTERVAL_MS > Date.now()) {
+        logger.debug(
+            "New session already forced with device " + sender + ":" + deviceKey +
+            " at " + lastNewSessionForced + ": not forcing another",
+        );
+        return;
```
Comment (uhoreg) [inline src/crypto/index.js]: I think the keyshare request should still be resent in this situation?

Comment (dbkr) [reply src/crypto/index.js]: I don't think so - we'll have resent all outstanding keyshare requests when the session was re-established, so any responses coming back on the old session will already have had their re-requests sent after establishment of the new session.

**Comment by uhoreg** [comment]
> I don't think so - we'll have resent all outstanding keyshare requests
> when the session was re-established, so any responses coming back on
> the old session will already have had their re-requests sent after
> establishment of the new session.

I'm not sure, but we can give it a go.  So, lgtm.

**Pull request merged**